README.md
# Audiomentations

[![Build status](https://img.shields.io/circleci/project/github/iver56/audiomentations/master.svg)](https://circleci.com/gh/iver56/audiomentations)
[![Code coverage](https://img.shields.io/codecov/c/github/iver56/audiomentations/master.svg)](https://codecov.io/gh/iver56/audiomentations)
[![Code Style: Black](https://img.shields.io/badge/code%20style-black-black.svg)](https://github.com/ambv/black)
[![Licence: MIT](https://img.shields.io/pypi/l/audiomentations)](https://github.com/iver56/audiomentations/blob/master/LICENSE)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7010042.svg)](https://doi.org/10.5281/zenodo.7010042)

A Python library for audio data augmentation. Inspired by
[albumentations](https://github.com/albu/albumentations). Useful for deep learning. Runs on
CPU. Supports mono audio and multichannel audio. Can be
integrated in training pipelines in e.g. Tensorflow/Keras or Pytorch. Has helped people get
world-class results in Kaggle competitions. Is used by companies making next-generation audio
products.

Need a Pytorch-specific alternative with GPU support? Check out [torch-audiomentations](https://github.com/asteroid-team/torch-audiomentations)!

# Setup

![Python version support](https://img.shields.io/pypi/pyversions/audiomentations)
[![PyPI version](https://img.shields.io/pypi/v/audiomentations.svg?style=flat)](https://pypi.org/project/audiomentations/)
[![Number of downloads from PyPI per month](https://img.shields.io/pypi/dm/audiomentations.svg?style=flat)](https://pypi.org/project/audiomentations/)

`pip install audiomentations`

# Usage example

```python
from audiomentations import Compose, AddGaussianNoise, TimeStretch, PitchShift, Shift
import numpy as np

augment = Compose([
    AddGaussianNoise(min_amplitude=0.001, max_amplitude=0.015, p=0.5),
    TimeStretch(min_rate=0.8, max_rate=1.25, p=0.5),
    PitchShift(min_semitones=-4, max_semitones=4, p=0.5),
    Shift(min_fraction=-0.5, max_fraction=0.5, p=0.5),
])

# Generate 2 seconds of dummy audio for the sake of example
samples = np.random.uniform(low=-0.2, high=0.2, size=(32000,)).astype(np.float32)

# Augment/transform/perturb the audio data
augmented_samples = augment(samples=samples, sample_rate=16000)
```

# Transforms

* AddBackgroundNoise
* AddGaussianNoise
* AddGaussianSNR
* ApplyImpulseResponse
* AddShortNoises
* AirAbsorption
* BandPassFilter
* BandStopFilter
* Clip
* ClippingDistortion
* FrequencyMask
* Gain
* GainTransition
* HighPassFilter
* HighShelfFilter
* LowPassFilter
* LowShelfFilter
* Mp3Compression
* Lambda
* Limiter
* LoudnessNormalization
* Normalize
* Padding
* PeakingFilter
* PitchShift
* PolarityInversion
* Resample
* Reverse
* RoomSimulator
* SevenBandParametricEQ
* Shift
* SpecChannelShuffle
* SpecFrequencyMask
* TanhDistortion
* TimeMask
* TimeStretch
* Trim

# Documentation

See [https://iver56.github.io/audiomentations/](https://iver56.github.io/audiomentations/)

# Acknowledgements

Thanks to [Nomono](https://nomono.co/) for backing audiomentations.

Thanks to [all contributors](https://github.com/iver56/audiomentations/graphs/contributors) who help improving audiomentations.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## [0.26.0] - 2022-08-19

### Added

* Add new transform `Lambda`. Thanks to Thanatoz-1.
* Add new transform `Limiter`. Thanks to pzelasko.

### Fixed

* Fix incorrect type hints in `RoomSimulator`
* Make `Shift` robust to different sample rate inputs when parameters are frozen

## [0.25.1] - 2022-06-15

### Fixed

* Fix a bug where `RoomSimulator` would treat an x value as if it was y, and vice versa

## [0.25.0] - 2022-05-30

### Added

* Add `AirAbsorption` transform
* Add mp4 to the list of recognized audio filename extensions

### Changed

* Guard against invalid params in `TimeMask`
* Emit `FutureWarning` instead of `UserWarning` in `Trim` and `ApplyImpulseResponse`
* Allow specifying a file path, a folder path, a list of files or a list of folders to
  `ApplyImpulseResponse`, `AddBackgroundNoise` and `AddShortNoises`. Previously only a path to a folder was allowed.

### Fixed

* Fix a bug with `noise_transform` in `AddBackgroundNoise` where some
  SNR calculations were done before the `noise_transform` was applied. This has sometimes
  led to incorrect SNR in the output. **This changes the behavior** of
  `AddBackgroundNoise` (when noise_transform is used).

### Removed

* Remove support for Python 3.6, as it is past its end of life already. RIP.

## [0.24.0] - 2022-03-18

### Added

* Add `SevenBandParametricEQ` transform
* Add optional `noise_transform` in `AddShortNoises`
* Add .aac and .aif to the list of recognized audio filename endings

### Changed

* Show warning if `top_db` and/or `p` in `Trim` are not specified because their default
  values will change in a future version

### Fixed

* Fix filter instability bug related to center freq above nyquist freq in `LowShelfFilter` and `HighShelfFilter`

## [0.23.0] - 2022-03-07

### Added

* Add `Padding` transform
* Add `RoomSimulator` transform for simulating shoebox rooms using `pyroomacoustics`
* Add parameter `signal_gain_in_db_during_noise` in `AddShortNoises`

### Changed

* Not specifying a value for `leave_length_unchanged` in `AddImpulseResponse` now emits
  a warning, as the default value will change from `False` to `True` in a future version.

### Removed

* Remove the deprecated `AddImpulseResponse` alias. Use `ApplyImpulseResponse` instead.
* Remove support for the legacy parameters `min_SNR` and `max_SNR` in `AddGaussianSNR`
* Remove useless default path value in `AddBackgroundNoise`, `AddShortNoises` and `ApplyImpulseResponse`

## [0.22.0] - 2022-02-18

### Added

* Implement `GainTransition`
* Add support for librosa 0.9
* Add support for stereo audio in `Mp3Compression`, `Resample` and `Trim`
* Add `"relative_to_whole_input"` option for `noise_rms` parameter in `AddShortNoises`
* Add optional `noise_transform` in `AddBackgroundNoise`

### Changed

* Improve speed of `PitchShift` by 6-18% when the input audio is stereo

### Removed

* Remove support for librosa<=0.7.2

## [0.21.0] - 2022-02-10

### Added

* Add support for multichannel audio in `ApplyImpulseResponse`, `BandPassFilter`, `HighPassFilter` and `LowPassFilter`
* Add `BandStopFilter` (similar to FrequencyMask, but with overhauled defaults and parameter randomization behavior), `PeakingFilter`, `LowShelfFilter` and `HighShelfFilter`
* Add parameter `add_all_noises_with_same_level` in `AddShortNoises`

### Changed

* Change `BandPassFilter`, `LowPassFilter`, `HighPassFilter`, to use scipy's butterworth
  filters instead of pydub. Now they have parametrized roll-off. Filters are now steeper
  than before by default - set `min_rolloff=6, max_rolloff=6` to get the old behavior.
  They also support zero-phase filtering now. And they're at least ~25x times faster than before!

### Removed

* Remove optional `wavio` dependency for audio loading

## [0.20.0] - 2021-11-18

### Added

* Implement `OneOf` and `SomeOf` for applying one of or some of many transforms. Transforms are randomly
  chosen every call. Inspired by augly. Thanks to Cangonin and iver56.
* Add a new argument `apply_to_children` (bool) in `randomize_parameters`,
  `freeze_parameters` and `unfreeze_parameters` in `Compose` and `SpecCompose`.

### Changed

* Insert three new parameters in `AddBackgroundNoise`: `noise_rms` (defaults to "relative", which is
  the old behavior), `min_absolute_rms_in_db` and `max_absolute_rms_in_db`. This **may be a breaking
  change** if you used `AddBackgroundNoise` with positional arguments in earlier versions of audiomentations!
  Please use keyword arguments to be on the safe side - it should be backwards compatible then.

### Fixed

* Remove global `pydub` import which was accidentally introduced in v0.18.0. `pydub` is
  considered an optional dependency and is imported only on demand now.

## [0.19.0] - 2021-10-18

### Added

* Implement `TanhDistortion`. Thanks to atamazian and iver56.
* Add a `noise_rms` parameter to `AddShortNoises`. It defaults to `relative`, which
  is the old behavior. `absolute` allows for adding loud noises to parts that are
  relatively silent in the input.

## [0.18.0] - 2021-08-05

### Added

* Implement `BandPassFilter`, `HighPassFilter`, `LowPassFilter` and `Reverse`. Thanks to atamazian.

## [0.17.0] - 2021-06-25

### Added

* Add a `fade` option in `Shift` for eliminating unwanted clicks
* Add support for 32-bit int wav loading with scipy>=1.6
* Add support for float64 wav files. However, the use of this format is discouraged,
  since float32 is more than enough for audio in most cases.
* Implement `Clip`. Thanks to atamazian.
* Add some parameter sanity checks in `AddGaussianNoise`
* Officially support librosa 0.8.1

### Changed

* Rename `AddImpulseResponse` to `ApplyImpulseResponse`. The former will still work for
  now, but give a warning.

demo/demo.py
import os
import random
from pathlib import Path

import numpy as np
import time
from scipy.io import wavfile
from tqdm import tqdm

from audiomentations import (
    AddGaussianNoise,
    TimeStretch,
    PitchShift,
    Shift,
    Normalize,
    FrequencyMask,
    TimeMask,
    AddGaussianSNR,
    Resample,
    ClippingDistortion,
    AddBackgroundNoise,
    AddShortNoises,
    PeakingFilter,
    PolarityInversion,
    Gain,
    Mp3Compression,
    LoudnessNormalization,
    Trim,
    LowPassFilter,
    LowShelfFilter,
    HighShelfFilter,
    HighPassFilter,
    BandPassFilter,
    ApplyImpulseResponse,
    Reverse,
    RoomSimulator,
    TanhDistortion,
    Compose,
    SomeOf,
    OneOf,
    BandStopFilter,
    GainTransition,
    Padding,
    AirAbsorption,
    Lambda,
)
from audiomentations.augmentations.limiter import Limiter
from audiomentations.augmentations.seven_band_parametric_eq import SevenBandParametricEQ
from audiomentations.core.audio_loading_utils import load_sound_file
from audiomentations.core.transforms_interface import (
    MultichannelAudioNotSupportedException,
)

DEMO_DIR = os.path.dirname(__file__)


class timer(object):
    """
    timer: A class used to measure the execution time of a block of code that is
    inside a "with" statement.

    Example:

    ```
    with timer("Count to 500000"):
        x = 0
        for i in range(500000):
            x += 1
        print(x)
    ```

    Will output:
    500000
    Count to 500000: 0.04 s

    Warning: The time resolution used here may be limited to 1 ms
    """

    def __init__(self, description="Execution time", verbose=False):
        self.description = description
        self.verbose = verbose
        self.execution_time = None

    def __enter__(self):
        self.t = time.time()
        return self

    def __exit__(self, type, value, traceback):
        self.execution_time = time.time() - self.t
        if self.verbose:
            print("{}: {:.3f} s".format(self.description, self.execution_time))


if __name__ == "__main__":
    """
    For each transformation, apply it to an example sound and write the transformed sounds to
    an output folder. Also crudely measure and print execution time.
    """
    output_dir = os.path.join(DEMO_DIR, "output")
    os.makedirs(output_dir, exist_ok=True)

    np.random.seed(420)
    random.seed(420)

    sound_file_paths = [
        Path(os.path.join(DEMO_DIR, "acoustic_guitar_0.wav")),
        Path(os.path.join(DEMO_DIR, "perfect-alley1.ogg")),
        Path(os.path.join(DEMO_DIR, "p286_011.wav")),
    ]

    transforms = [
        {
            "instance": AddBackgroundNoise(
                sounds_path=os.path.join(DEMO_DIR, "background_noises"), p=1.0
            ),
            "num_runs": 5,
            "name": "AddBackgroundNoiseRelative",
        },
        {
            "instance": AddBackgroundNoise(
                sounds_path=os.path.join(DEMO_DIR, "background_noises"),
                noise_rms="absolute",
                min_absolute_rms_in_db=-30,
                max_absolute_rms_in_db=-10,
                p=1.0,
            ),
            "num_runs": 5,
            "name": "AddBackgroundNoiseAbsolute",
        },
        {
            "instance": AddBackgroundNoise(
                sounds_path=os.path.join(DEMO_DIR, "background_noises"),
                min_snr_in_db=2,
                max_snr_in_db=4,
                noise_transform=Reverse(p=1.0),
                p=1.0,
            ),
            "num_runs": 5,
            "name": "AddBackgroundNoiseWithTransform",
        },
        {
            "instance": AddGaussianNoise(
                min_amplitude=0.001, max_amplitude=0.015, p=1.0
            ),
            "num_runs": 5,
        },
        {
            "instance": AddGaussianSNR(p=1.0),
            "num_runs": 5,
            "name": "AddGaussianSNR",
        },
        {
            "instance": ApplyImpulseResponse(
                p=1.0,
                ir_path=os.path.join(DEMO_DIR, "ir"),
                leave_length_unchanged=False,
            ),
            "num_runs": 1,
            "name": "ApplyImpulseResponseWithTail",
        },
        {
            "instance": ApplyImpulseResponse(
                p=1.0, ir_path=os.path.join(DEMO_DIR, "ir"), leave_length_unchanged=True
            ),
            "num_runs": 1,
            "name": "ApplyImpulseResponseLeaveLengthUnchanged",
        },
        {
            "instance": AddShortNoises(
                sounds_path=os.path.join(DEMO_DIR, "short_noises"),
                min_absolute_noise_rms_db=-30,
                max_absolute_noise_rms_db=-10,
                noise_rms="absolute",
                min_time_between_sounds=2.0,
                max_time_between_sounds=4.0,
                burst_probability=0.4,
                min_pause_factor_during_burst=0.01,
                max_pause_factor_during_burst=0.95,
                min_fade_in_time=0.005,
                max_fade_in_time=0.08,
                min_fade_out_time=0.01,
                max_fade_out_time=0.1,
                p=1.0,
            ),
            "num_runs": 5,
            "name": "AddShortNoisesAbsolute",
        },
        {
            "instance": AddShortNoises(
                sounds_path=os.path.join(DEMO_DIR, "short_noises"),
                min_snr_in_db=0,
                max_snr_in_db=8,
                noise_rms="relative",
                min_time_between_sounds=2.0,
                max_time_between_sounds=4.0,
                burst_probability=0.4,
                min_pause_factor_during_burst=0.01,
                max_pause_factor_during_burst=0.95,
                min_fade_in_time=0.005,
                max_fade_in_time=0.08,
                min_fade_out_time=0.01,
                max_fade_out_time=0.1,
                p=1.0,
            ),
            "num_runs": 5,
            "name": "AddShortNoisesRelative",
        },
        {
            "instance": AddShortNoises(
                sounds_path=os.path.join(DEMO_DIR, "short_noises"),
                min_snr_in_db=0,
                max_snr_in_db=8,
                noise_rms="relative",
                min_time_between_sounds=2.0,
                max_time_between_sounds=4.0,
                burst_probability=0.4,
                min_pause_factor_during_burst=0.01,
                max_pause_factor_during_burst=0.95,
                min_fade_in_time=0.005,
                max_fade_in_time=0.08,
                min_fade_out_time=0.01,
                max_fade_out_time=0.1,
                signal_gain_in_db_during_noise=-100.0,
                p=1.0,
            ),
            "num_runs": 5,
            "name": "AddShortNoisesWithSignalGain",
        },
        {
            "instance": AddShortNoises(
                sounds_path=os.path.join(DEMO_DIR, "short_noises"),
                min_snr_in_db=0,
                max_snr_in_db=8,
                noise_rms="relative",
                min_time_between_sounds=1.0,
                max_time_between_sounds=2.0,
                burst_probability=0.4,
                min_pause_factor_during_burst=0.01,
                max_pause_factor_during_burst=0.95,
                min_fade_in_time=0.005,
                max_fade_in_time=0.08,
                min_fade_out_time=0.01,
                max_fade_out_time=0.1,
                noise_transform=OneOf(
                    [
                        TanhDistortion(min_distortion=0.8, max_distortion=0.99),
                        RoomSimulator(
                            calculation_mode="rt60",
                            min_target_rt60=2.2,
                            max_target_rt60=3.0,
                            leave_length_unchanged=False,
                        ),
                    ]
                ),
                p=1.0,
            ),
            "num_runs": 5,
            "name": "AddShortNoisesWithNoiseTransform",
        },
        {"instance": BandPassFilter(p=1.0), "num_runs": 5},
        {"instance": BandStopFilter(p=1.0), "num_runs": 5},
        {"instance": ClippingDistortion(p=1.0), "num_runs": 5},
        {
            "instance": FrequencyMask(
                min_frequency_band=0.5, max_frequency_band=0.6, p=1.0
            ),
            "num_runs": 5,
        },
        {"instance": Gain(min_gain_in_db=-6, max_gain_in_db=6, p=1.0), "num_runs": 5},
        {"instance": GainTransition(p=1.0), "num_runs": 5},
        {"instance": HighPassFilter(p=1.0), "num_runs": 5},
        {"instance": HighShelfFilter(p=1.0), "num_runs": 5},
        {"instance": LowPassFilter(p=1.0), "num_runs": 5},
        {"instance": LowShelfFilter(p=1.0), "num_runs": 5},
        {
            "instance": PitchShift(min_semitones=-4, max_semitones=4, p=1.0),
            "num_runs": 5,
        },
        {
            "instance": Lambda(
                transform=lambda samples, sample_rate: samples - 0.2,
                p=1.0,
            ),
            "num_runs": 1,
        },
        {"instance": Limiter(p=1.0), "num_runs": 5},
        {"instance": LoudnessNormalization(p=1.0), "num_runs": 5},
        {
            "instance": Mp3Compression(backend="lameenc", p=1.0),
            "num_runs": 5,
            "name": "Mp3CompressionLameenc",
        },
        {
            "instance": Mp3Compression(backend="pydub", p=1.0),
            "num_runs": 5,
            "name": "Mp3CompressionPydub",
        },
        {"instance": Normalize(p=1.0), "num_runs": 1},
        {
            "instance": Padding(mode="silence", pad_section="end", p=1.0),
            "num_runs": 5,
            "name": "PaddingSilenceEnd",
        },
        {
            "instance": Padding(mode="wrap", pad_section="end", p=1.0),
            "num_runs": 5,
            "name": "PaddingWrapEnd",
        },
        {
            "instance": Padding(mode="reflect", pad_section="end", p=1.0),
            "num_runs": 5,
            "name": "PaddingReflectEnd",
        },
        {
            "instance": Padding(mode="silence", pad_section="start", p=1.0),
            "num_runs": 5,
            "name": "PaddingSilenceStart",
        },
        {
            "instance": Padding(mode="wrap", pad_section="start", p=1.0),
            "num_runs": 5,
            "name": "PaddingWrapStart",
        },
        {"instance": PeakingFilter(p=1.0), "num_runs": 5},
        {"instance": PolarityInversion(p=1.0), "num_runs": 1},
        {"instance": Resample(p=1.0), "num_runs": 5},
        {"instance": Reverse(p=1.0), "num_runs": 1},
        {
            "instance": Compose(
                [RoomSimulator(p=1.0, leave_length_unchanged=True), Normalize(p=1.0)]
            ),
            "num_runs": 5,
            "name": "RoomSimulator",
        },
        {
            "instance": Compose([SevenBandParametricEQ(p=1.0), Normalize(p=1.0)]),
            "num_runs": 5,
            "name": "SevenBandParametricEQ",
        },
        {
            "instance": Shift(min_fraction=-0.5, max_fraction=0.5, fade=False, p=1.0),
            "num_runs": 5,
            "name": "ShiftWithoutFade",
        },
        {
            "instance": Shift(min_fraction=-0.5, max_fraction=0.5, fade=True, p=1.0),
            "num_runs": 5,
            "name": "ShiftWithShortFade",
        },
        {
            "instance": Shift(
                min_fraction=-0.5,
                max_fraction=0.5,
                rollover=False,
                fade=True,
                fade_duration=0.3,
                p=1.0,
            ),
            "num_runs": 5,
            "name": "ShiftWithoutRolloverWithLongFade",
        },
        {"instance": TanhDistortion(p=1.0), "num_runs": 5},
        {"instance": TimeMask(p=1.0), "num_runs": 5},
        {"instance": TimeStretch(min_rate=0.8, max_rate=1.25, p=1.0), "num_runs": 5},
        {"instance": Trim(p=1.0), "num_runs": 1},
        {
            "instance": Compose(
                [
                    AddGaussianNoise(min_amplitude=0.001, max_amplitude=0.015, p=0.5),
                    SomeOf(
                        (0, 2),
                        [
                            TimeStretch(min_rate=0.8, max_rate=1.25, p=1.0),
                            PitchShift(min_semitones=-4, max_semitones=4, p=1.0),
                        ],
                    ),
                    Shift(min_fraction=-0.5, max_fraction=0.5, p=0.5),
                    OneOf([TanhDistortion(p=1.0), ClippingDistortion(p=1.0)], p=0.25),
                ]
            ),
            "num_runs": 10,
            "name": "BigCompose",
        },
        {"instance": AirAbsorption(p=1.0), "num_runs": 5},
    ]

    for sound_file_path in sound_file_paths:
        samples, sample_rate = load_sound_file(
            sound_file_path, sample_rate=None, mono=False
        )
        if len(samples.shape) == 2 and samples.shape[0] > samples.shape[1]:
            samples = samples.transpose()

        print(
            "Transforming {} with shape {}".format(
                sound_file_path.name, str(samples.shape)
            )
        )
        execution_times = {}

        for transform in tqdm(transforms):
            augmenter = transform["instance"]
            run_name = (
                transform.get("name")
                if transform.get("name")
                else transform["instance"].__class__.__name__
            )
            execution_times[run_name] = []
            for i in range(transform["num_runs"]):
                output_file_path = os.path.join(
                    output_dir,
                    "{}_{}_{:03d}.wav".format(sound_file_path.stem, run_name, i),
                )
                try:
                    with timer() as t:
                        augmented_samples = augmenter(
                            samples=samples, sample_rate=sample_rate
                        )
                    execution_times[run_name].append(t.execution_time)

                    if len(augmented_samples.shape) == 2:
                        augmented_samples = augmented_samples.transpose()

                    wavfile.write(
                        output_file_path, rate=sample_rate, data=augmented_samples
                    )
                except MultichannelAudioNotSupportedException as e:
                    print(e)

        for run_name in execution_times:
            if len(execution_times[run_name]) > 1:
                print(
                    "{:<32} {:.3f} s (std: {:.3f} s)".format(
                        run_name,
                        np.mean(execution_times[run_name]),
                        np.std(execution_times[run_name]),
                    )
                )
            else:
                print(
                    "{:<32} {:.3f} s".format(
                        run_name, np.mean(execution_times[run_name])
                    )
                )
